infra/util/Crc.hpp

```diff
@@ -0,0 +1,170 @@
+#ifndef INFRA_CRC_HPP
+#define INFRA_CRC_HPP
+
+#include "infra/util/ByteRange.hpp"
+#include <climits>
+#include <cstddef>
+#include <cstdint>
+
+namespace infra
+{
+    template<typename CRC_TYPE>
+    constexpr CRC_TYPE PolyReverse(CRC_TYPE Polynomial);
+
+    template<typename CRC_TYPE, CRC_TYPE Polynomial, CRC_TYPE InitValue = 0, CRC_TYPE FinalXor = 0, bool ReflectInput = false, bool ReflectOutput = false>
+    class Crc
+    {
+        static_assert(ReflectInput == ReflectOutput, "Currently, non-matching ReflectInput and ReflectOutput is not supported");
+
+    public:
+        Crc() = default;
+        ~Crc() = default;
+
+        void Update(uint8_t input);
+        void Update(ConstByteRange bytes);
+        CRC_TYPE Result() const;
+        void Reset();
+
+    private:
+        class Table
+        {
+        public:
+            constexpr Table();
+            CRC_TYPE operator[](int i) const;
+
+        private:
+            CRC_TYPE table[256];
+        };
+
+        template<typename TYPE>
+        static constexpr TYPE Reflect(TYPE value);
+
+        static constexpr uint8_t bitWidth = sizeof(CRC_TYPE) * CHAR_BIT;
+        static constexpr bool bitWidthEquals8 = bitWidth == 8;
+        static constexpr uint8_t shift = bitWidth - 8;
+        static inline constexpr Table table{};
+
+        CRC_TYPE crc = InitValue;
+    };
+
+    using Crc8Maxim = Crc<uint8_t, 0x31, 0, 0, true, true>;
+    using Crc8Sensirion = Crc<uint8_t, 0x31, 0xff, 0, false, false>;
+    using Crc16Modbus = Crc<uint16_t, 0x8005, 0xffff, 0, true, true>;
+    using Crc16Xmodem = Crc<uint16_t, 0x1021, 0x0>;
+    using Crc16CcittFalse = Crc<uint16_t, 0x1021, 0xffff>;
+    using Crc64Ecma = Crc<uint64_t, 0x42F0E1EBA9EA3693, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, true, true>;
+
```

Comment:
Could you also add a ISO CRC32 typedef?
```cpp
    using Crc32 = Crc<uint32_t, 0x04C11DB7>;

```

Reply:
Used correct initvalue 0xffffffff. Added testcase